File: backend/app/api/georef_suggest.py
"""Automatic plan alignment — POST /api/georef/suggest.

The client sends the rendered plan bitmap (its own «Deckung prüfen» snapshot), the object
coordinate and the calibration-derived metres-per-pixel; the server segments the sheet,
fetches OSM building rings through the station's Overpass proxy and returns a similarity fit
as TWO reference pairs (`kind: "auto"`) plus its score. The client seeds «Deckung prüfen»
with them — the operator reviews, adjusts and only then accepts. Nothing is stored here.

The 200 answer is an NDJSON STREAM, so the busy card can show the real phases instead of one
opaque spinner over a 3–15 s wait: a `{"step": "osm"|"match"}` line as each phase begins, then
exactly one of `{"result": {…SuggestResponse}}` or `{"error": "<German sentence>"}` (an error
after the phases have begun cannot change the status line any more — it travels in-band).
Everything checkable BEFORE the stream begins still answers a plain HTTP error.

The heavy CV dependencies live in the optional `georef` group (`uv sync --extra georef`);
without them this endpoint answers 503 and the app falls back to «Punkte selbst setzen».
Editor-auth: a suggestion is a step in authoring a georeference, which is editor work.
"""

import json
import logging
import math
import time
from collections.abc import AsyncIterator

import anyio
from fastapi import APIRouter, HTTPException, UploadFile, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from .. import overpass as overpass_client
from ..auth.dependencies import CurrentEditor

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/georef", tags=["georef"])

# An A3 sheet at 150 dpi is ~2 MB as JPEG; anything past this is not a plan snapshot.
_MAX_UPLOAD = 8 * 1024 * 1024

# ⚠️ The DECODED size needs its own bound — the JPEG cap does not give one (8 MB decodes to
# far more), and the resize below multiplies it. A real client raster is ≤1755 px long side
# (< 2.2 MP); past this the request is not a plan snapshot, whatever it compressed to.
_MAX_DECODED_PX = 6_000_000
# …and the normalization must never be allowed to build a monster either: the resample target
# is capped so an adversarial (or simply wrong) mPerPx cannot turn one editor request into a
# multi-gigabyte allocation that OOM-kills the backend for everyone.
_MAX_RESIZED_SIDE = 6000

# The matcher is a 3–15 s CPU burn per request and a thread cannot be cancelled once started —
# its own small limiter keeps a burst of retries from pegging every core and starving the PDF
# renderers, which share the default thread pool.
_match_limiter = anyio.CapacityLimiter(2)


def _load_matcher():
    """The lazy heavy-deps import, as a SEAM: the 503-fail-closed path is exactly what holds on
    a production image without the `georef` extra, so it must be testable from an environment
    where the extra IS installed — the test monkeypatches this to raise."""
    from .. import georef_suggest

    return georef_suggest


# Overpass answers for one object change on OSM's timescale, not the operator's: cache them per
# rounded coordinate so a re-run after «Anpassen»/a retry does not re-ask the public mirrors
# (which rate-limit), and serve the last good answer when every mirror momentarily fails.
_OSM_TTL_S = 600
_osm_cache: dict[str, tuple[float, dict]] = {}


async def _osm_around(lng: float, lat: float, radius_m: float) -> dict:
    # the radius is part of the identity: a Modul-1 attempt needs the WIDE box, and must not be
    # handed the narrow one a Modul-2 attempt on the same object just cached
    key = f"{round(lng, 4)},{round(lat, 4)},{round(radius_m)}"
    now = time.monotonic()
    hit = _osm_cache.get(key)
    if hit and now - hit[0] < _OSM_TTL_S:
        return hit[1]
    dlat = radius_m / 111_320
    dlng = radius_m / (111_320 * math.cos(math.radians(lat)))
    bbox = f"{lat - dlat},{lng - dlng},{lat + dlat},{lng + dlng}"
    try:
        data = await overpass_client.fetch_buildings(overpass_client.BUILDINGS_QUERY.format(bbox=bbox))
    except Exception:
        # stale beats a 502 — buildings do not move. The entry may be well past the TTL (it
        # only ever leaves by LRU eviction), which is still the better answer than nothing.
        if hit:
            return hit[1]
        raise
    _osm_cache[key] = (now, data)
    if len(_osm_cache) > 64:
        del _osm_cache[min(_osm_cache, key=lambda k: _osm_cache[k][0])]
    return data


class _Pt(BaseModel):
    x: float
    y: float


class _LngLat(BaseModel):
    lng: float
    lat: float


class SuggestPair(BaseModel):
    plan: _Pt
    lngLat: _LngLat  # noqa: N815
    kind: str


class SuggestResponse(BaseModel):
    """`found: false` is a normal answer (segmentation found nothing, or the best fit scored
    past the ceiling) — the surface then offers the manual point flow, not an error toast.
    `confident: false` marks the band between cutoff and ceiling: a pose worth reviewing,
    surfaced with the «Deckung nachprüfen» warning instead of the plain proposal head."""

    found: bool
    confident: bool | None = None
    pairs: list[SuggestPair] = []
    rotationDeg: float | None = None  # noqa: N815
    score: float | None = None
    coverage: float | None = None
    seconds: float


@router.post("/suggest")
async def suggest_alignment(
    _user: CurrentEditor,
    image: UploadFile,
    lng: float,
    lat: float,
    mPerPx: float,  # noqa: N803 — camelCase mirrors the client wire format
    template: str = "m2",
) -> StreamingResponse:
    try:
        matcher = _load_matcher()
    except ImportError as exc:
        raise HTTPException(
            status.HTTP_503_SERVICE_UNAVAILABLE,
            "Automatische Ausrichtung ist auf diesem Server nicht installiert",
        ) from exc

    # mPerPx tops out well under 0.5 for every real template (1:2500 @ 150 dpi = 0.42); the
    # bound is generous but keeps the resize factor — mPerPx / TUNED_M_PER_PX — small
    if not (-180 <= lng <= 180 and -90 <= lat <= 90) or not (0 < mPerPx < 1) or template not in ("m1", "m2"):
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, "ungültige Parameter")
    if not overpass_client.mirrors():
        raise HTTPException(status.HTTP_503_SERVICE_UNAVAILABLE, "Overpass ist nicht konfiguriert")

    raw = await image.read(_MAX_UPLOAD + 1)
    if len(raw) > _MAX_UPLOAD:
        raise HTTPException(status.HTTP_413_CONTENT_TOO_LARGE, "Bild zu gross")
    if not raw:
        # cv2.imdecode ASSERTS on an empty buffer instead of returning None — refuse it here
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, "Bild nicht lesbar")

    started = time.monotonic()

    def run(osm: dict) -> SuggestResponse:
        import cv2
        import numpy as np

        img = cv2.imdecode(np.frombuffer(raw, np.uint8), cv2.IMREAD_COLOR)
        if img is None:
            raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, "Bild nicht lesbar")
        if img.shape[0] * img.shape[1] > _MAX_DECODED_PX:
            raise HTTPException(status.HTTP_413_CONTENT_TOO_LARGE, "Bild zu gross")
        # Normalize to what the segmentation kernels were tuned at. Sheets in the evaluated
        # ground-resolution band get the page-size rule (A4 @ 150 dpi ≈ 1755 px long side — a
        # phone's ~900 px bake would otherwise lose its grey fills to the 9×9 opening); a
        # COARSER sheet (Modul 1 at 1:2500 ≈ 0.42 m/px) is instead resampled toward the tuned
        # ground resolution, or a 10 m house stays smaller than the kernel however big the page.
        h, w = img.shape[:2]
        m_per_px = mPerPx
        factor = (m_per_px / matcher.TUNED_M_PER_PX) if m_per_px > 0.20 else 1755 / max(h, w)
        factor = min(factor, _MAX_RESIZED_SIDE / max(h, w))  # bounded, whatever mPerPx claimed
        if abs(factor - 1) > 0.05:
            img = cv2.resize(
                img,
                (max(1, round(w * factor)), max(1, round(h * factor))),
                interpolation=cv2.INTER_AREA if factor < 1 else cv2.INTER_CUBIC,
            )
            m_per_px = m_per_px / factor
        rings = matcher.rings_from_overpass(osm, lng, lat)
        seconds = lambda: round(time.monotonic() - started, 2)  # noqa: E731
        try:
            s = matcher.suggest(img, m_per_px, rings, template)
        except ValueError as exc:
            logger.info("georef suggest: no fit (%s)", exc)
            return SuggestResponse(found=False, seconds=seconds())
        ceiling = matcher.SCORE_CEILING_M1 if template == "m1" else matcher.SCORE_CEILING
        if s.score > ceiling:
            logger.info("georef suggest: rejected by score %.2f (template %s)", s.score, template)
            return SuggestResponse(found=False, score=round(s.score, 2), seconds=seconds())
        pairs = matcher.suggestion_pairs(s, img.shape[1], img.shape[0], lng, lat)
        return SuggestResponse(
            found=True,
            # M1 has no independently validated confidence threshold. A lower union score
            # fixes compound ranking; it must not silently promote this template's claim.
            confident=template != "m1" and s.score <= matcher.SCORE_CUTOFF,
            pairs=[SuggestPair(**p) for p in pairs],
            rotationDeg=round(s.rotation_deg, 2),
            score=round(s.score, 2),
            coverage=round(s.coverage, 3),
            seconds=seconds(),
        )

    # the OSM box scales with the sheet's printed ground extent (a 1:2500 Modul 1 covers far
    # more than the evaluated ±190 m); decoded lazily just for the dimensions is wasteful, so
    # approximate the long side from the upload — the client sends ≤1755 px rasters
    radius = matcher.reference_radius_m(mPerPx, 1755)

    async def stream() -> AsyncIterator[str]:
        yield json.dumps({"step": "osm"}) + "\n"
        try:
            osm = await _osm_around(lng, lat, radius)
        except Exception:  # noqa: BLE001 — headers are sent; ANY fetch failure must go in-band
            yield json.dumps({"error": "Overpass nicht erreichbar"}) + "\n"
            return
        yield json.dumps({"step": "match"}) + "\n"
        try:
            # CPU-bound for 3–15 s — off the event loop, and through the matcher's OWN limiter
            result = await anyio.to_thread.run_sync(run, osm, limiter=_match_limiter)
        except HTTPException as exc:
            yield json.dumps({"error": str(exc.detail)}) + "\n"
            return
        except Exception:
            logger.exception("georef suggest failed")
            yield json.dumps({"error": "Automatische Ausrichtung fehlgeschlagen"}) + "\n"
            return
        yield json.dumps({"result": result.model_dump()}) + "\n"

    return StreamingResponse(stream(), media_type="application/x-ndjson")

File: src/lib/georefSuggest.ts
/** «Automatisch ausrichten» — the client half of the CV alignment suggestion.
 *
 *  The heavy lifting happens in the backend (POST /api/georef/suggest, app/georef_suggest.py):
 *  this module renders the plan into the same ≤1800 px bake «Deckung prüfen» uses, sends it with
 *  the object coordinate and the calibration-derived metres-per-pixel, and hands back the fit as
 *  two `kind: 'auto'` reference pairs. Those seed the mode's PROPOSAL review (lib/georefMode ·
 *  startGeorefProposal) — nothing is stored until the operator presses «Übernehmen».
 *
 *  The matcher is template-specific (the Modul-2 sheet family it was evaluated on), needs the
 *  printed scale via the station calibration, and an anchor coordinate to fetch OSM building
 *  rings around — `georefSuggestEligible` is the one place those preconditions live.
 */
import { ApiError, apiUploadRaw } from './api'
import { planMatcherImage, planPrintedMPerU } from '../components/PdfViewport'
import type { GeoPt, GeorefPair } from './georef'

/** The phases the busy card can show — 'render' is the client's own bake, the other two are
 *  the server's NDJSON progress lines (app/api/georef_suggest). */
export type GeorefSuggestStep = 'render' | 'osm' | 'match'

/** What the server hands back for a usable fit. `score`/`coverage` are the matcher's own
 *  internal quality reads (lower score = better; neither is metres or a probability);
 *  `confident: false` = the pose is in the review-carefully band — the coverage view leads
 *  with «Deckung nachprüfen» instead of the plain proposal head. */
export interface GeorefSuggestion {
  pairs: GeorefPair[]
  confident: boolean
  rotationDeg: number
  score: number
  coverage: number
  seconds: number
}

interface SuggestWire {
  found: boolean
  confident?: boolean | null
  pairs?: { plan: { x: number; y: number }; lngLat: { lng: number; lat: number } }[]
  rotationDeg?: number
  score?: number
  coverage?: number
  seconds: number
}

/** May «Automatisch ausrichten» be offered for this sheet? Every MODULE sheet with an
 *  object/incident coordinate to anchor the OSM reference box gets the chooser (decided
 *  08.09.2026 after the Modul-2-only gate proved undiscoverable twice): the matcher was
 *  evaluated on the Modul-2 template — on other templates it simply runs and comes back
 *  «kein Vorschlag» honestly (Modul 1 failed 4/4 in the experiment), which the operator can
 *  judge; a hidden capability they cannot. The SCALE is resolved at request time — the
 *  sheet's own printed «1:NNN» first, the calibration as fallback — so it is no precondition. */
export function georefSuggestEligible(planId: string, anchor: GeoPt | null | undefined): boolean {
  return /^modul\d/.test(planId) && !!anchor
}

/** What asking the matcher can come back with, short of a transport/server failure. */
export type GeorefSuggestOutcome =
  | { kind: 'fit'; suggestion: GeorefSuggestion }
  /** ran, found nothing confident — a normal outcome, the manual flow is the way on */
  | { kind: 'none' }
  /** never ran: neither a printed «1:NNN» on the sheet nor a calibration says the scale, and a
   *  fixed-scale matcher fed a guessed scale returns confident-LOOKING wrong poses */
  | { kind: 'noScale' }

/**
 * Ask the server for an alignment suggestion. Rejects on transport/server failure (offline,
 * 503 while the extra is not installed); everything else is a typed outcome.
 *
 * ⚠️ The expected scale comes from the sheet's own printed «1:NNN» (planPrintedMPerU) FIRST
 * and only falls back to the calibration: the station-default calibration is one number for
 * every sheet, but Oberwil's own Modul 2s mix 1:750 and 1:1000 — and the fixed-scale ICP fed a
 * scale 25 % off finds a WRONG pose that still clears the score cutoff (measured on the
 * Gymnasium sheet, 08.09.2026).
 *
 * Takes 3–14 s server-side — callers show a busy state and must tolerate the operator
 * navigating away meanwhile.
 */
export async function requestGeorefSuggestion(opts: {
  /** the plan's own PDF url (`PlanDocument.imageUrl`) — rendered here, not on the server */
  planUrl: string
  anchor: GeoPt
  /** calibration fallback: metres per aspect-corrected normalized unit (`PlanScale.mPerU`) */
  mPerU?: number
  /** which sheet TEMPLATE the matcher should segment — 'm1' for the Modul-1 family (red
   *  Einsatzobjekt compound, 1:2000+ extents), 'm2' (default) for everything else */
  template?: 'm1' | 'm2'
}, onStep?: (step: GeorefSuggestStep) => void): Promise<GeorefSuggestOutcome> {
  onStep?.('render')
  const printed = await planPrintedMPerU(opts.planUrl).catch(() => null)
  const mPerU = printed ?? opts.mPerU
  if (!mPerU || !(mPerU > 0)) return { kind: 'noScale' }
  // reuses the sheet's resident bake whenever possible (PdfViewport · planMatcherImage) —
  // «Plan rendern» is then an encode, not a second full pdf.js rasterization
  const blob = await planMatcherImage(opts.planUrl)
  const bmp = await createImageBitmap(blob)
  const mPerPx = mPerU / bmp.height
  bmp.close()
  const form = new FormData()
  form.append('image', blob, 'plan.jpg')
  const q = `lng=${opts.anchor.lng}&lat=${opts.anchor.lat}&mPerPx=${mPerPx}&template=${opts.template ?? 'm2'}`
  const res = await readSuggestStream(await apiUploadRaw(`/api/georef/suggest?${q}`, form), onStep)
  if (!res.found || !res.pairs || res.pairs.length < 2) return { kind: 'none' }
  return {
    kind: 'fit',
    suggestion: {
      pairs: res.pairs.map((p) => ({ plan: p.plan, lngLat: p.lngLat, kind: 'auto' as const })),
      confident: res.confident !== false,
      rotationDeg: res.rotationDeg ?? 0,
      score: res.score ?? 0,
      coverage: res.coverage ?? 0,
      seconds: res.seconds,
    },
  }
}

/** Read the endpoint's NDJSON answer: progress lines feed `onStep`, an in-band `{error}`
 *  becomes the same ApiError a plain failure would be, `{result}` is the payload. */
async function readSuggestStream(res: Response, onStep?: (step: GeorefSuggestStep) => void): Promise<SuggestWire> {
  const handle = (line: string): SuggestWire | null => {
    if (!line.trim()) return null
    const msg = JSON.parse(line) as { step?: string; error?: string; result?: SuggestWire }
    if (msg.step === 'osm' || msg.step === 'match') onStep?.(msg.step)
    if (msg.error) throw new ApiError(502, msg.error)
    return msg.result ?? null
  }
  const reader = res.body?.getReader()
  if (!reader) {
    // no streaming reader — the whole body arrived buffered; the progress is lost, the answer is not
    for (const line of (await res.text()).split('\n')) {
      const result = handle(line)
      if (result) return result
    }
    throw new ApiError(0, 'georef suggest: stream ended without a result')
  }
  const dec = new TextDecoder()
  let buf = ''
  for (;;) {
    const { done, value } = await reader.read()
    if (value) buf += dec.decode(value, { stream: !done })
    let nl = buf.indexOf('\n')
    while (nl >= 0) {
      const line = buf.slice(0, nl)
      buf = buf.slice(nl + 1)
      const result = handle(line)
      if (result) return result
      nl = buf.indexOf('\n')
    }
    if (done) {
      const result = handle(buf)
      if (result) return result
      throw new ApiError(0, 'georef suggest: stream ended without a result')
    }
  }
}
